src/main.rs

```diff
@@ -99,12 +101,22 @@ fn get_gps(config: Rc<Config>) -> Box<GPS> {
 
         Err(e) => {
             match e.kind() {
-                ::std::io::ErrorKind::NotFound => println!("{}", e),
+                ::std::io::ErrorKind::NotFound => match GNSS::new(config.clone()) {
```

Comment:
nice! However, this makes code a bit hard to follow (due to indentation). Also, it'll get even harder to read, when/if someone adds support for another device type. So I'd suggest separating out the discovery part from RS232 module first (in a seperate  patch) and then adding GNSS support. That way this code will not have to change for addition of new devices at all and will be much more readable.

I'm sorry I didn't think of this before but as I mentioned, I've not looked at this for a while. If you don't have the time to do this, I understand but please do add a FIXME comment here then, how to improve this code (basically what I wrote above).

Reply:
Oh and keep in mind that when I wrote this project, I was completely new to Rust. :)

Reply:
Ok I'll add a FIXME as I'm short on time and skill with rust.